banyand/measure/measure_topn.go

```diff
@@ -0,0 +1,568 @@
+	streamingApi "github.com/apache/skywalking-banyandb/pkg/flow/streaming/api"
+	"github.com/apache/skywalking-banyandb/pkg/logger"
+	"github.com/apache/skywalking-banyandb/pkg/partition"
+	pbv1 "github.com/apache/skywalking-banyandb/pkg/pb/v1"
+	"github.com/apache/skywalking-banyandb/pkg/timestamp"
+)
+
+const (
+	timeBucketFormat = "200601021504"
+	TopNTagFamily    = "__topN__"
+)
+
+var (
+	_ bus.MessageListener = (*topNProcessCallback)(nil)
+	_ io.Closer           = (*topNStreamingProcessor)(nil)
+	_ io.Closer           = (*topNProcessorManager)(nil)
+	_ streamingApi.Sink   = (*topNStreamingProcessor)(nil)
+
+	TopNValueFieldSpec = &databasev1.FieldSpec{
+		Name:              "value",
+		FieldType:         databasev1.FieldType_FIELD_TYPE_INT,
+		EncodingMethod:    databasev1.EncodingMethod_ENCODING_METHOD_GORILLA,
+		CompressionMethod: databasev1.CompressionMethod_COMPRESSION_METHOD_ZSTD,
+	}
+)
+
+type (
+	measureProcessorFilter func(request *measurev1.DataPointValue) bool
+	measureProcessorMapper func(request *measurev1.DataPointValue) api.Data
+)
+
+type topNStreamingProcessor struct {
+	l                *logger.Logger
+	shardNum         uint32
+	interval         time.Duration
+	topNSchema       *databasev1.TopNAggregation
+	databaseSupplier tsdb.Supplier
+	src              chan interface{}
+	in               chan interface{}
+	errCh            <-chan error
+	streamingFlow    api.Flow
+}
+
+func (t *topNStreamingProcessor) In() chan<- interface{} {
+	return t.in
+}
+
+func (t *topNStreamingProcessor) Setup(ctx context.Context) error {
+	go t.run(ctx)
+	return nil
+}
+
+func (t *topNStreamingProcessor) run(ctx context.Context) {
+	for {
+		select {
+		case item, open := <-t.in:
+			if !open {
+				return
+			}
+			if record, ok := item.(api.StreamRecord); ok {
+				if err := t.writeStreamRecord(record); err != nil {
+					t.l.Err(err).Msg("fail to write stream record")
+				}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (t *topNStreamingProcessor) Teardown(ctx context.Context) error {
+	return nil
+}
+
+func (t *topNStreamingProcessor) Close() error {
+	close(t.src)
+	return nil
+}
+
+func (t *topNStreamingProcessor) writeStreamRecord(record api.StreamRecord) error {
+	tuples, ok := record.Data().([]*streaming.Tuple2)
+	if !ok {
+		return errors.New("invalid data type")
+	}
+	// eventTime is the start time of a timeWindow
+	eventTime := time.UnixMilli(record.TimestampMillis())
+	// down-sampling to a time bucket as measure ID
+	timeBucket := t.downSampleTimeBucket(eventTime)
+	var err error
+	for rankNum, tuple := range tuples {
+		fieldValue := tuple.V1.(int64)
+		data := tuple.V2.(api.Data)
+		err = multierr.Append(err, t.writeData(eventTime, timeBucket, fieldValue, data, rankNum))
+	}
+	return err
+}
+
+func (t *topNStreamingProcessor) writeData(eventTime time.Time, timeBucket string, fieldValue int64, data api.Data, rankNum int) error {
+	// HACK: use rankNumber as the nanoseconds to avoid overwrite items with same entity
```

Comment:
To avoid data overwrite, I currently add rankNumber as nanoseconds. For example, we want to save a snapshot of the TopN ranking at a given moment,

| Rank | GroupValue | FieldValue |
| ----- | ----------- | ---- |
| 1  | `Entity_1` |  10 |
| 2 | `Entity_2` | 5 |
| 3 | `Entity_1` |  2 |

Since we use `GroupValue`(s) as the entity, if we store (1) and (3) using the same eventTime, only the one saved later can be read in the query.

An alternative would be add rankNumber as a part of the `entity` but probably cause worse compression ratio? @hanahmily any suggestion?

Reply:
Now the rankNumber is used as a part of the `Entity` for consistency.